src/test/model.test.ts

```diff
@@ -30,20 +41,45 @@ class M extends Model<M> {
   }
 }
 
-
 class N extends Model<N> {
   x: number
   y: string
 
   keys() {
+    return [indexBy(X).exact(this.x), indexBy(Y).exact(this.y)]
+  }
+}
+
+class O extends Model<O> {
```

Comment:
I suspect a test model here with variable number of shadow key sets would be pretty nice in covering that particular use-case as well.